pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,290 @@
+package ftp
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/jlaffaye/ftp"
+)
+
+// Conn struct embeds the *ftp.ServerConn returned by ftp server on successful connection.
+type Conn struct {
+	*ftp.ServerConn
+}
+
+// Retr wraps the ftp retrieve method to return a ftpResponse interface type.
+func (c *Conn) Retr(path string) (ftpResponse, error) {
+	return c.ServerConn.Retr(path)
+}
+
+func (c *Conn) RetrFrom(path string, offset uint64) (ftpResponse, error) {
+	return c.ServerConn.RetrFrom(path, offset)
+}
+
+var (
+	errEmptyFilename  = errors.New("filename cannot be empty")
+	errEmptyPath      = errors.New("file/directory path cannot be empty")
+	errEmptyDirectory = errors.New("directory name cannot be empty")
+	errInvalidArg     = errors.New("invalid filename/directory")
+	transferComplete  = "226 Transfer complete."
+)
+
+// ftpFileSystem represents a file system interface over FTP.
+type ftpFileSystem struct {
+	*ftpFile
+	conn   ServerConn
+	config *Config
+	logger Logger
+}
+
+// Config represents the FTP configuration.
+type Config struct {
+	Host      string // FTP server hostname
+	User      string // FTP username
+	Password  string // FTP password
+	Port      int    // FTP port
+	RemoteDir string // Remote directory path. Base Path for all FTP Operations.
+}
+
+// New initializes a new instance of ftpFileSystem with provided configuration.
+func New(config *Config) FileSystemProvider {
+	return &ftpFileSystem{config: config}
+}
+
+// UseLogger sets the logger interface for the FTP file system.
+func (f *ftpFileSystem) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		f.logger = l
+	}
+}
+
+// UseMetrics sets the metrics for the ftpFileSystem client which asserts the Metrics interface.
+// Currently not implemented.
+func (*ftpFileSystem) UseMetrics(_ interface{}) {
+
+}
+
+// Connect establishes a connection to the FTP server and logs in.
+func (f *ftpFileSystem) Connect() {
+	ftpServer := fmt.Sprintf("%v:%v", f.config.Host, f.config.Port)
+
+	const dialTimeout = 5 * time.Second
+
+	conn, err := ftp.Dial(ftpServer, ftp.DialWithTimeout(dialTimeout))
+	if err != nil {
+		f.logger.Errorf("Connection Failed Error: %v", err)
+		return
+	}
+
+	f.conn = &Conn{conn}
+
+	f.logger.Logf("Connection Success Connected to FTP Server : %v", ftpServer)
+
+	err = conn.Login(f.config.User, f.config.Password)
+	if err != nil {
+		f.logger.Errorf("Login Failed Error: %v", err)
+		return
+	}
+
+	f.logger.Logf("Login Success Current remote location: %q", f.config.RemoteDir)
+}
+
+// Create creates an empty file on the FTP server.
+func (f *ftpFileSystem) Create(name string) (File, error) {
+	if name == "" {
+		f.logger.Errorf("Create_File Failed Provide a valid filename: %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	emptyReader := new(bytes.Buffer)
+
+	filePath := fmt.Sprintf("%s/%s", f.config.RemoteDir, name)
```

Comment:
Please use `filepath.Join`

https://pkg.go.dev/path/filepath#Join

Reply:
Please, _do not use `filepath.Join` for remote file paths. This causes an incompatibility when a Windows client talks to a *nix server.

`path.Join` should be used instead.

Reply:
Oh that's right. I missed that